### PR TITLE
fix serialization for large buffers

### PR DESCRIPTION
**Pull request by fitouch**
Fixes `Maximum call stack size exceeded` error which triggers when passing large buffers.

### CODE REVIEW

**Comment by amark** [comment]
@sirpy you able to take a peek?

File: src/serializeBinary.ts [hunk 1]
```diff
@@ -9,31 +9,37 @@ export async function parse(text: string): Promise<any> {
...
-
 const ArrayBufferSerializer: Serializer<ArrayBuffer, string> = {
   id: "ArrayBuffer",
   isType: (o: any) => o instanceof ArrayBuffer,
 
   // from https://developers.google.com/web/updates/2012/06/How-to-convert-ArrayBuffer-to-and-from-String
   // modified to use Int8Array so that we can hold odd number of bytes
   toObject: async (ab: ArrayBuffer) => {
-    return String.fromCharCode.apply(null, new Int8Array(ab));
+    return new Int8Array(ab).reduce(function (data, byte) {
```
Comment (sirpy) [inline src/serializeBinary.ts]: @fitouch according to https://www.measurethat.net/Benchmarks/Show/1219/23/arraybuffer-to-base64-string
using apply is order of magnitude faster then going char by char

I suggest you split the `ab` buffer into chunks each of MAX_BUFFER_LENGTH
and do the charCode.apply on each of these buffers and concat them

**Review by sirpy — CHANGES_REQUESTED** [review]

File: src/serializeBinary.ts (hunk 1 above)
Comment (fitouch) [reply src/serializeBinary.ts]: @sirpy here's a new commit based on your suggestion, thank you

File: src/serializeBinary.ts [hunk 2]
```diff
@@ -37,9 +37,17 @@ const ArrayBufferSerializer: Serializer<ArrayBuffer, string> = {
   // from https://developers.google.com/web/updates/2012/06/How-to-convert-ArrayBuffer-to-and-from-String
   // modified to use Int8Array so that we can hold odd number of bytes
   toObject: async (ab: ArrayBuffer) => {
-    return new Int8Array(ab).reduce(function (data, byte) {
-      return data + String.fromCharCode(byte);
-    }, "");
+    const bytes = new Int8Array(ab);
+    const MAX_BUFFER_LENGTH = 8 * 1024;
```
Comment (sirpy) [inline src/serializeBinary.ts]: why 8k? whats the buffer sizes you have issues with?
also please add code comment about the bug you encountered and why this buffer splitting strategy is required to solve it

Comment (fitouch) [reply src/serializeBinary.ts]: Sorry, 64k is the buffer sizes I have issue with. I just re-commit again. This is for large base64 string.

File: src/serializeBinary.ts [hunk 3]
```diff
@@ -9,31 +9,45 @@ export async function parse(text: string): Promise<any> {
...
   // from https://developers.google.com/web/updates/2012/06/How-to-convert-ArrayBuffer-to-and-from-String
   // modified to use Int8Array so that we can hold odd number of bytes
   toObject: async (ab: ArrayBuffer) => {
-    return String.fromCharCode.apply(null, new Int8Array(ab));
+    const bytes = new Int8Array(ab);
+    const MAX_BUFFER_LENGTH = 64 * 1024;
+    if (bytes.length <= MAX_BUFFER_LENGTH)
+      return String.fromCharCode.apply(null, bytes);
+    let str = "";
+    for (let i = 0; i < bytes.length; i += MAX_BUFFER_LENGTH)
```
Comment (sirpy) [inline src/serializeBinary.ts:46]:
```suggestion
    // Fixes Maximum call stack size exceeded error which triggers when passing large buffers.
    for (let i = 0; i < bytes.length; i += MAX_BUFFER_LENGTH)
```

**Comment by sirpy** [comment]
@fitouch merged and released

**Comment by fitouch** [comment]
@sirpy thank you

**Pull request merged**